feat: Document BigQuery exports (#6707)
PostHog/posthog.com: contents/docs/cdp/batch-exports/bigquery.md

2. Create a [key](https://cloud.google.com/iam/docs/keys-create-delete#creating) for the Service Account you created in the previous step.
3. Save the key file as JSON to upload it when configuring a batch export.

![Create JSON private key](../../images/docs/batch-exports/bigquery/create-private-key-json.png)

4. Create a role which has only the specific permissions the batch export requires (listed below), or use the built in `BigQuery Data Owner` and `BigQuery Job User` roles. If you create a custom role, you will need:
   * `bigquery.datasets.get`
   * `bigquery.jobs.create`
   * `bigquery.tables.create`
   * `bigquery.tables.get`
   * `bigquery.tables.list`
   * `bigquery.tables.updateData`

![Create custom role for batch exports](../../images/docs/batch-exports/bigquery/create-role.png)

5. Grant the Service Account access to run jobs in your Google Cloud project. This can be done by granting the `BigQuery Jobs User` role on your project, or with the role we have just created in the previous step.

Navigate to IAM and click on Grant Access to arrive at this screen:

![Project grant](../../images/docs/batch-exports/bigquery/project-grant.png)

> In the screenshot above, we have used a custom role named `Testing PostHog BatchExports` with the permissions listed in the previous step.

Author: This one is already applied 👍

6. Create a dataset within a BigQuery project (ours is called `BatchExports`, but any name will do).

> 7. Follow the instructions on [granting access to a dataset in BigQuery](https://cloud.google.com/bigquery/docs/control-access-to-resources-iam#grant_access_to_a_dataset):
> 
> ![Sharing dataset](../../images/docs/batch-exports/bigquery/dataset-sharing.png)
> ![Add principal](../../images/docs/batch-exports/bigquery/dataset-add-principal.png)
> 
> Use the Sharing and Add Principal buttons to share your dataset with your new Service Account created in step 1.
> 
> Grant either the `BigQuery Data Owner` role, or your custom role created in step 4 to the Service Account. In the screenshot below, we have used a custom role named `Testing PostHog BatchExports`.

Reviewer: Restructuring this to make it easier to follow. 
```suggestion
7. Use the Sharing and Add Principal buttons to share your dataset with your new Service Account created in step 1. Grant either the `BigQuery Data Owner` role or your custom role created in step 4 to the Service Account. Read the full instructions on [granting access to the dataset in BigQuery](https://cloud.google.com/bigquery/docs/control-access-to-resources-iam#grant_access_to_a_dataset) if unclear.

![Sharing dataset](../../images/docs/batch-exports/bigquery/dataset-sharing.png)
![Add principal](../../images/docs/batch-exports/bigquery/dataset-add-principal.png)

> In the screenshot below, we grant a custom role named `Testing PostHog BatchExports` access to the `BatchExports` data set.
```

Author: Not quite: The role only dictates the permissions (actions that the principal can do), but doesn't receive any grants. The Service Account (the principal) is the one getting assigned the permissions dictated in the `Testing PostHog BatchExports` role to operate on the BatchExports dataset. 

I may be making things more confusing, I definitely get lost with the right terminology.

Reviewer: How about:

> In the screenshot below, we grant our Service Account access to the `BatchExports` data set and assign the `Testing PostHog BatchExports` role permissions for it.

Updated my suggestion with the change.

Author: Yeah, that sounds better 👍

Author: I can't commit the suggestion from here as it's marked as out of date, but I'll make a new commit locally with the same changes.

Author: Done! Thank you!

![Sharing dataset](../../images/docs/batch-exports/bigquery/dataset-grant-access.png)

> 8. All done! After completing these steps you can create a BigQuery batch export and your data will start flowing from PostHog to BigQuery.

Reviewer: ```suggestion
8. All done! After completing these steps you can create a BigQuery [batch export in PostHog](https://app.posthog.com/project/apps?tab=batch_exports) and your data will start flowing from PostHog to BigQuery.
```

| set_once              | `STRING`    | A JSON object with any person properties sent with the `$set_once` field                                            |
| distinct_id           | `STRING`    | The `distinct_id` of the user who sent the event                                                                    |
| team_id               | `STRING`    | The `team_id` for the event                                                                                         |
| ip                    | `STRING`    | The IP address that was sent with the event                                                                         |
| site_url              | `STRING`    | The $current_url property of the event. This field has been kept for backwards compatibility and will be deprecated |
| timestamp             | `TIMESTAMP` | The timestamp associated with an event                                                                              |
| bq_ingested_timestamp | `TIMESTAMP` | The timestamp when the event was sent to BigQuery                                                                   |

## Creating the batch export

> 1. Navigate to the exports page in your PostHog instance (Quick links if you use [PostHog Cloud US](https://app.posthog.com/batch_exports) or [PostHog Cloud EU](https://eu.posthog.com/batch_exports)).

Reviewer: ```suggestion
1. Navigate to the exports page in your PostHog instance (Quick links if you use [PostHog Cloud US](https://app.posthog.com/project/apps?tab=batch_exports) or [PostHog Cloud EU](https://eu.posthog.com/project/apps?tab=batch_exports)).
```

2. Click "Create export workflow".
3. Select **BigQuery** as the batch export destination.
4. Fill in the necessary [configuration details](#bigquery-configuration).
5. Finalize the creation by clicking on "Create".
6. Done! The batch export will schedule its first run on the start of the next period.

## BigQuery configuration

Configuring a batch export targeting BigQuery requires the following BigQuery-specific configuration values:

> * **Table ID:** The ID of a BigQuery table where to export the data. This is not the fully-qualified name of a table, so omit the dataset and project IDs. For example for the fully-qualified table name `project-123:dataset:MyExportTable` only `MyExporTable` is to be used as the table ID.

Reviewer: ```suggestion
* **Table ID:** The ID of the destination BigQuery table. This is not the fully-qualified name of a table, so omit the dataset and project IDs. For example for the fully-qualified table name `project-123:dataset:MyExportTable`, use only `MyExportTable` as the table ID.

```

> * **Dataset ID:** The ID of a BigQuery dataset where the table to export the data is located. Only the dataset ID is required, so omit the project ID. For example for the dataset `project-123:my-dataset` only `my-dataset` is to be used as the dataset ID.

Reviewer: Is `exported data table` correct? The wording is a bit confusing here, maybe I got it wrong too. 
```suggestion
* **Dataset ID:** The ID of a BigQuery dataset where the exported data table is located. Only the dataset ID is required, so omit the project ID. For example for the dataset `project-123:my-dataset`, only use `my-dataset` as the dataset ID.

```

Author: A google cloud project can contain one or more BigQuery datasets. We are batch exporting data to one table in one of these datasets. So maybe, something more like this?:
> * **Dataset ID:** The ID of the BigQuery dataset which contains the exported data table.

Reviewer: That's good

Author: I've  tweaked it to say 
> ... which contains the destination table
Just to link it with the previous configuration item (Table ID) as that one is defined as the "... destination BigQuery table".

Hopefully this works!

1. Navigate to the exports page in your PostHog instance (Quick links if you use [PostHog Cloud US](https://app.posthog.com/batch_exports) or [PostHog Cloud EU](https://eu.posthog.com/batch_exports)).
2. Click "Create export workflow".
3. Select **BigQuery** as the batch export destination.
4. Fill in the necessary [configuration details](#bigquery-configuration).
5. Finalize the creation by clicking on "Create".
6. Done! The batch export will schedule its first run on the start of the next period.

## BigQuery configuration

Configuring a batch export targeting BigQuery requires the following BigQuery-specific configuration values:
* **Table ID:** The ID of a BigQuery table where to export the data. This is not the fully-qualified name of a table, so omit the dataset and project IDs. For example for the fully-qualified table name `project-123:dataset:MyExportTable` only `MyExporTable` is to be used as the table ID.
* **Dataset ID:** The ID of a BigQuery dataset where the table to export the data is located. Only the dataset ID is required, so omit the project ID. For example for the dataset `project-123:my-dataset` only `my-dataset` is to be used as the dataset ID.

> * **Dataset ID:** The ID of a BigQuery dataset where the table to export the data is located. Only the dataset ID is required, so omit the project ID. For example for the dataset `project-123:my-dataset` only `my-dataset` is to be used as the dataset ID.

Reviewer 2: dup row - remove

* **Google Cloud JSON key file:** In order to access your BigQuery instance, we use a Service Account. When creating one, a JSON key file will be generated. That JSON file needs to be uploaded here. See [here](#setting-up-bigquery-access) for more information.

> ### Setting up BigQuery access

Reviewer 2: I'd move this to be the first thing as that's what users need to do first

Author: Yeah, makes sense. I've moved this first. Thanks

| uuid                  | `STRING`    | The unique ID of the event within PostHog                                        |
| event                 | `STRING`    | The name of the event that was sent                                              |
| properties            | `STRING`    | A JSON object with all the properties sent along with an event                   |
| elements              | `STRING`    | A string of elements surrounding an [autocaptured](/docs/data/autocapture) event |
| set                   | `STRING`    | A JSON object with any person properties sent with the `$set` field              |
| set_once              | `STRING`    | A JSON object with any person properties sent with the `$set_once` field         |
| distinct_id           | `STRING`    | The `distinct_id` of the user who sent the event                                 |
| team_id               | `STRING`    | The `team_id` for the event                                                      |
| ip                    | `STRING`    | The IP address that was sent with the event                                      |

> | site_url              | `STRING`    | The $current_url property of the event                                           |

Reviewer 2: Can we drop this everywhere? Context for what it is https://posthog.slack.com/archives/C0374DA782U/p1687955284804389 it's essentially only ever used for when webhooks link to the person etc, so it likely shouldn't be in batch exports anywhere

Author: We can. However, the docs reflect how the system looks right now. Once we drop the field, we can remove this from here.

Reviewer 2: any reason not to drop the field now?

Author: The code is not here.

Author: In other words, I don't know what this question is doing in a PR about docs. Happy to review a PR to remove this over at https://github.com/PostHog/posthog.

Author: The fair thing to ask in _this PR_ is to warn users the field will be deprecated, which I have done in my latest commit.

Reviewer 2: I don't impose restrictions like that on my PR comments & I prefer if others don't either. 
I either didn't notice or review your PR that added site_url, so I commented about it the first time I saw/noticed it - referring to the best actual solution.
